### ISSUE TITLE
Stopping criteria

### ISSUE THREAD

**Issue by oxinabox** [FluxML/Flux.jl#227]
It came up in #131 and #132,
I think we should have a number of predefined stopping callbacks,
So that you feed the training data in as an an infintity stream of epochs.
And you have stop based on criteria.
The ones I am thinking are:

- [ ]  `atol` stop when the loss is below some point (obv. 0 as default)
- [ ]  `rtol` stop when the decrease in loss from one epoch to the next is below some point (obv. 0 as default)
- [ ] Max time spent
- [ ] Max Epochs

 `atol` and `rtol` could be parameterised to use different loss functions,
including different datasets. Early stopping can be achieved by using a development dataset instead of the training dataset.

We could have more; e.g. incorporating some from of hysteresis to for looking at improvement over n-epochs, like some kind of `rⁿtol` but  that can be later.

Of course the user should also still be able to define their own custom methods.

**Comment by ToucheSir** [comment]
Looks like #1545 has captured most of this. The rest is arguably better suited for an external package like https://github.com/JuliaAI/EarlyStopping.jl.

**Comment by darsnack** [comment]
Yeah I think this is closed by #1545.
